File: python/get_equal_substrings_within_budget.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        """time complexity: O(n)"""
        left: int = 0
        curr_cost: int = 0
        result: int = 0
        for right in range(len(t)):
            curr_cost += abs(ord(s[right]) - ord(t[right]))
            while curr_cost > maxCost:
                curr_cost -= abs(ord(s[left]) - ord(t[left]))
                left += 1
            result = max(result, right - left + 1)
        return result


def test_equalSubstring_case_1():
    # arrange
    s: str = "abcd"
    t: str = "bcdf"
    maxCost: int = 3
    expected: int = 3

    # act
    actual = Solution().equalSubstring(s, t, maxCost)

    # assert
    assert expected == actual


def test_equalSubstring_case_2():
    # arrange
    s: str = "abcd"
    t: str = "cdef"
    maxCost: int = 3
    expected: int = 1

    # act
    actual = Solution().equalSubstring(s, t, maxCost)

    # assert
    assert expected == actual


def test_equalSubstring_case_3():
    # arrange
    s: str = "abcd"
    t: str = "acde"
    maxCost: int = 0
    expected: int = 1

    # act
    actual = Solution().equalSubstring(s, t, maxCost)

    # assert
    assert expected == actual

File: rust/src/code/get_equal_substrings_within_budget.rs
struct Solution;

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn equal_substring(s: String, t: String, max_cost: i32) -> i32 {
        use std::cmp::max;

        let s: Vec<char> = s.chars().collect();
        let t: Vec<char> = t.chars().collect();
        let mut l: usize = 0;
        let mut result: i32 = 0;
        let mut curr_cost: i32 = 0;
        for r in 0..s.len() {
            curr_cost += ((s[r] as i8 - t[r] as i8) as i32).abs();
            while curr_cost > max_cost {
                curr_cost -= ((s[l] as i8 - t[l] as i8) as i32).abs();
                l += 1;
            }
            if l <= r {
                result = max(result, (r - l + 1) as i32);
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_equal_substring_case_1() {
        // arrange
        let s: String = "abcd".to_string();
        let t: String = "bcdf".to_string();
        let maxCost: i32 = 3;
        let expected: i32 = 3;

        // act
        let actual = Solution::equal_substring(s, t, maxCost);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_equal_substring_case_2() {
        // arrange
        let s: String = "abcd".to_string();
        let t: String = "cdef".to_string();
        let maxCost: i32 = 3;
        let expected: i32 = 1;

        // act
        let actual = Solution::equal_substring(s, t, maxCost);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_equal_substring_case_3() {
        // arrange
        let s: String = "abcd".to_string();
        let t: String = "acde".to_string();
        let maxCost: i32 = 0;
        let expected: i32 = 1;

        // act
        let actual = Solution::equal_substring(s, t, maxCost);

        // assert
        assert_eq!(expected, actual);
    }
}
